exercises/practice/list-ops/test_list_ops.c

```diff
@@ -92,248 +133,234 @@ static list_element_t fold_divide(list_element_t element,
 
 static void test_append_empty_lists(void)
 {
-   list_t *list1 = new_list(0, NULL);
-   list_t *list2 = new_list(0, NULL);
+   list = new_list(0, NULL);
+   list2 = new_list(0, NULL);
 
-   list_t *actual = append_list(list1, list2);
+   actual = append_list(list, list2);
    check_lists_match(0, NULL, actual);
 
-   delete_list(list1);
-   delete_list(list2);
-   delete_list(actual);
+   call_delete_list(3, &list, &list2, &actual);
```

Comment:
If I'm a student learning C and I see this, I'm not sure that it will be clear to me what's happening. This is less verbose, but for a student is this an improvement over deleting each of these and setting to `NULL`?

Reply:
That the students read the tests to learn C vs. just to get the function signatures is something I hadn't considered. I can say, as a student of C myself, when I saw the three commands in a row that only differ by the argument, my first thought was "why not send them all at once into a single function that can loop through them?" And that set me onto a variadic function. If not a variadic function, I would think to at least send them into a function that would call `delete_lst` and set to NULL instead of doing it inline over and over. Then again, if they are learning C from reading the tests, then this is how they can learn about variadic functions.

Reply:
I wasn't approaching this from the perspective of using the tests as an opportunity for a student to learn more C. My concern is instead that the student should be able to read the test to understand what is being tested and not have to learn, for example, what a variadic function is in order to reach the learning objective of the exercise. The test code should not get in the way of the student completing the exercise.

I can speak from my experience in other tracks on this. Many tracks have "clever" tests that are structured to be as DRY as possible in order to make them easier to maintain. But, as a student, this often makes it unclear what is even being tested. There may be an argument for using this approach on "real" tests (ie: tests for production code), but as a learner that is still coming to grips with the language, this almost invariably serves to obfuscate. Instead of being able to focus on the learning objectives of the exercise, I also have to wade through the complexity of the tests to understand what is expected of my code.

Reply:
```javascript
delete_list(list1);
list1 = NULL;
delete_list(list2);
list2 = NULL;
delete_list(actual);
actual = NULL;
```

doesn't seem [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) to me, but it may be simpler for a novice to follow.

As it is now, if students develop their solutions in the browser, they won't see the memory leaks happening, but if they develop locally they will see memory leaks if they run `make memcheck`. I don't know how many of them do that instead of running `make test`. I know I do, and the first time I saw memory leaks for a failed test when I thought I had handled it properly, it was confusing, until I realized it was the test itself which was leaking memory. So I think it should be fixed, even if we simplify the fix. I think it can be embarrassing to Exercism if we don't.

Reply:
Our last comments crossed. I can revert to the simpler fixes and remove the variadics..

Reply:
Does this seem okay, or do you think it's still too complex?

```c
static void call_delete_list(list_t **list)
{
   delete_list(*list);
   *list = NULL;
}
// code snipped
static void test_append_empty_lists(void)
{
   list = new_list(0, NULL);
   list2 = new_list(0, NULL);

   actual = append_list(list, list2);
   check_lists_match(0, NULL, actual);

   call_delete_list(&list);
   call_delete_list(&list2);
   call_delete_list(&actual);
}
```

Reply:
By "too complex" I mean is the indirection too confusing for a novice. To your point, I remember a particular exercise on another track which generated some complaints right after the v3 launch. The implementation was empty and a few students found it too hard to figure out what was required from the tests.

Reply:
> So I think it should be fixed, even if we simplify the fix. I think it can be embarrassing to Exercism if we don't.

I agree. My point is just that we need to make sure that the fix isn't worse for a student than the issue we're fixing. Your last proposal looks good to me. It's cleaner without relying upon anything that will look like black magic to a beginner.

Reply:
So, to make sure I understand what my last proposal is, the teardown will be this

```c
void tearDown(void)
{
   release(3, &list, &list2, &actual);
   if (error_message) {
      free(error_message);
      error_message = NULL;
   }
}
```

and the tests will be something like this

```c
   call_delete_list(&list);
   call_delete_list(&list2);
   call_delete_list(&actual);
```

Sorry if I got it wrong. I should know what my last proposal is, but since they're spit between two threads, I've befuddled myself!

Reply:
As the `tearDown()` now does the cleaning up, and `delete_list()` is not an explicit part of the spec, do we even need to delete the lists within the test functions?
They could be removed completely and just let `tearDown()` handle it in both the success case and the error case.